client/tso_client.go

```diff
@@ -242,136 +149,104 @@ func (c *tsoMcsClient) GetTSOAllocatorServingAddrByDCLocation(dcLocation string)
-func (c *tsoMcsClient) GetOrCreateGRPCConn(addr string) (*grpc.ClientConn, error) {
-	return grpcutil.GetOrCreateGRPCConn(c.ctx, &c.clientConns, addr, c.tlsCfg, c.option.gRPCDialOptions...)
+// AddTSOAllocatorServingAddrSwitchedCallback adds callbacks which will be called
+// when any global/local tso allocator service endpoint is switched.
+func (c *tsoClient) AddTSOAllocatorServingAddrSwitchedCallback(callbacks ...func()) {
+	c.tsoAllocServingAddrSwitchedCallback = append(c.tsoAllocServingAddrSwitchedCallback, callbacks...)
 }
 
-// ScheduleCheckMemberChanged is used to trigger a check to see if there is any change in ervice endpoints.
-func (c *tsoMcsClient) ScheduleCheckMemberChanged() {
-	select {
-	case c.checkMembershipCh <- struct{}{}:
-	default:
+func (c *tsoClient) updateTSOAllocatorServingAddrs(allocatorMap map[string]string) error {
+	if len(allocatorMap) == 0 {
+		return nil
 	}
-}
-
-// Immediately checkif there is any membership change among the primary/secondaries in
-// a primary/secondy configured cluster.
-func (c *tsoMcsClient) CheckMemberChanged() error {
-	return c.updateMember()
-}
-
-// AddServingAddrSwitchedCallback adds callbacks which will be called when the primary in
-// a primary/secondary configured cluster is switched.
-func (c *tsoMcsClient) AddServingAddrSwitchedCallback(callbacks ...func()) {
-	c.primarySwitchedCallbacks = append(c.primarySwitchedCallbacks, callbacks...)
-}
 
-// AddServiceAddrsSwitchedCallback adds callbacks which will be called when any primary/secondary
-// in a primary/secondary configured cluster is changed.
-func (c *tsoMcsClient) AddServiceAddrsSwitchedCallback(callbacks ...func()) {
-	c.membersChangedCallbacks = append(c.membersChangedCallbacks, callbacks...)
-}
+	updated := false
 
-// AddTSOAllocatorServingAddrSwitchedCallback adds callbacks which will be called
-// when any keyspace group tso allocator primary is switched.
-func (c *tsoMcsClient) AddTSOAllocatorServingAddrSwitchedCallback(callbacks ...func()) {
-	c.tsoAllocatorLeaderSwitchedCallback = append(c.tsoAllocatorLeaderSwitchedCallback, callbacks...)
-}
-
-// getPrimaryAddr returns the primary address.
-func (c *tsoMcsClient) getPrimaryAddr() string {
-	primaryAddr := c.primary.Load()
-	if primaryAddr == nil {
-		return ""
+	// Switch to the new one
+	for dcLocation, addr := range allocatorMap {
+		if len(addr) == 0 {
+			continue
+		}
+		oldAddr, exist := c.GetTSOAllocatorServingAddrByDCLocation(dcLocation)
+		if exist && addr == oldAddr {
+			continue
+		}
+		updated = true
+		if _, err := c.svcDiscovery.GetOrCreateGRPCConn(addr); err != nil {
+			log.Warn("[pd[tso]] failed to connect dc tso allocator serving address",
+				zap.String("dc-location", dcLocation),
+				zap.String("serving-address", addr),
+				errs.ZapError(err))
+			return err
+		}
+		c.tsoAllocators.Store(dcLocation, addr)
+		log.Info("[pd(tso)] switch dc tso allocatorimary",
```

Comment:
```suggestion
		log.Info("[pd(tso)] switch dc tso allocator primary",
```

Reply:
BTW, in PD only mode, do we need to mention the concept of primary?

Reply:
good catch! changed to "switch dc tso allocator serving address"

Reply:
No, in the tso components which are shared by both pd or tso modes, like tso_client and tso_dispatcher, I intentionally use neutral work like "serving address or endpoint" or use statement like "... followers in a quorum-based cluster or secondaries in a primary/secondary configured cluster."  This place, I just walk through the files, seems to be only place (if I didn't miss anything) that uses the wrong word.